packages/frontpage/lib/auth-sign-in.ts

```diff
@@ -0,0 +1,164 @@
+import {
+  calculatePKCECodeChallenge,
+  discoveryRequest,
+  generateKeyPair,
+  generateRandomCodeVerifier,
+  generateRandomState,
+  processDiscoveryResponse,
+  pushedAuthorizationRequest,
+} from "oauth4webapi";
+import {
+  getClientMetadata,
+  getClientPrivateKey,
+  getOauthClientOptions,
+  oauthProtectedMetadataRequest,
+} from "./auth";
+import { getDidFromHandleOrDid } from "./data/atproto/identity";
+import { oauthParResponseSchema } from "@atproto/oauth-types";
+import { db } from "./db";
+import * as schema from "./schema";
+import { redirect } from "next/navigation";
+
+export async function signIn(handle: string) {
+  const did = await getDidFromHandleOrDid(handle);
+  if (!did) {
+    return {
+      error: "DID_NOT_FOUND" as const,
+    };
+  }
+
+  const meta = await oauthProtectedMetadataRequest(did);
+  if ("error" in meta) {
+    return meta;
+  }
+
+  const authServerUrl = meta.data.authorization_servers?.[0];
+  if (!authServerUrl) {
+    return {
+      error: "NO_AUTH_SERVER" as const,
+    };
+  }
+
+  // TODO: Cache this
```

Comment:
Couldn't cache it in both node and edge environments because unstable_cache didn't work in edge runtime. 

Would be annoying to solve so just raised it for now and will revisit later https://github.com/vercel/next.js/issues/70085